RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.h

```diff
@@ -0,0 +1,194 @@
+
+  tensorflow::GraphDef* graphDef_HB_;
+  tensorflow::Session* tf_session_HB_;
+
+  tensorflow::GraphDef* graphDef_HLIP_;
+  tensorflow::Session* tf_session_HLIP_;
+
+  tensorflow::GraphDef* graphDef_HLMuS_;
+  tensorflow::Session* tf_session_HLMuS_;
+
+  /// DNN metadata
+  const std::string dnnMetadataPath_;
+  pt::ptree metadata;
+
+  /// Create seeds without hits on a given layer (TOB or TEC)
+  void makeSeedsWithoutHits(const GeometricSearchDet& layer,
+                            const TrajectoryStateOnSurface& tsos,
+                            const Propagator& propagatorAlong,
+                            edm::ESHandle<Chi2MeasurementEstimatorBase>& estimator,
+                            double errorSF,
+                            unsigned int& hitlessSeedsMade,
+                            unsigned int& numSeedsMade,
+                            std::vector<TrajectorySeed>& out) const;
+
+  /// Find hits on a given layer (TOB or TEC) and create seeds from updated TSOS with hit
+  void makeSeedsFromHits(const GeometricSearchDet& layer,
+                         const TrajectoryStateOnSurface& tsos,
+                         const Propagator& propagatorAlong,
+                         edm::ESHandle<Chi2MeasurementEstimatorBase>& estimator,
+                         edm::Handle<MeasurementTrackerEvent>& measurementTracker,
+                         unsigned int& hitSeedsMade,
+                         unsigned int& numSeedsMade,
+                         unsigned int& layerCount,
+                         std::vector<TrajectorySeed>& out) const;
+
+  /// Similar to makeSeedsFromHits, but seed is created only if there are compatible hits on two adjacent layers
+  void makeSeedsFromHitDoublets(const GeometricSearchDet& layer,
+                                const TrajectoryStateOnSurface& tsos,
+                                const Propagator& propagatorAlong,
+                                edm::ESHandle<Chi2MeasurementEstimatorBase>& estimator,
+                                edm::Handle<MeasurementTrackerEvent>& measurementTracker,
+                                edm::ESHandle<NavigationSchool> navSchool,
+                                unsigned int& hitDoubletSeedsMade,
+                                unsigned int& numSeedsMade,
+                                unsigned int& layerCount,
+                                std::vector<TrajectorySeed>& out) const;
+
+  /// Calculate the dynamic error SF by analysing the L2
+  double calculateSFFromL2(const reco::TrackRef track) const;
+
+  /// Dictionary of inputs for DNN
+  std::map<std::string, float> getFeatureMap(reco::TrackRef l2,
+                                             const TrajectoryStateOnSurface& tsos_IP,
+                                             const TrajectoryStateOnSurface& tsos_MuS) const;
+
+  /// Evaluate DNN classifier
+  std::tuple<int, int, int, float, bool> evaluateClassifier(std::map<std::string, float> feature_map,
```

Comment:
- I think that `struct` with clearly named member data is more clear than the unnamed tuples where it's easy to mixup the variables
- shouldn't `feature_map` and `session` be const here?

Reply:
from https://github.com/cms-sw/cmssw/pull/35237/files#r712600540
- shouldn't feature_map and session be const here?
```suggestion
  std::tuple<int, int, int, float, bool> evaluateClassifier(const std::map<std::string, float>& feature_map,
```

if there is a strong desire to stay with `std::tuple` (instead of a more readable struct with named members), please at least document the parameters of the `tuple` here and in the `evaluateRegressor` declaration

Reply:
Done!